django/contrib/sessions/backends/base.py

```diff
@@ -147,24 +156,60 @@ def is_empty(self):
         except AttributeError:
             return True
 
+    def _get_session_key_part(self, session_key):
+        "Return a string of the session_key key."
+        try:
+            algorithm, key = session_key.split(SESSION_KEY_DELIMITER)
+            if algorithm == SESSION_HASHING_ALGORITHM:
+                return key
+            return None
+        except ValueError:
+            return None
+
+    def _is_hashed(self, session_key):
+        "Return True when the session_key is hashed in the backend."
+        try:
+            return bool(self._get_session_key_part(session_key))
+        except IndexError:
+            return False
+
+    def get_session_key_hash(self, session_key):
+        """
+        Return a hash of the session key to be stored in the db. We want a fast
+        hashing function to minimize the impact on performance. We don't have to
+        worry about speed and salting since the session_key is already high entropy.
+        """
+        if session_key is not None:
+            if self._is_hashed(session_key) or settings.SESSION_REQUIRE_KEY_HASH:
+                key = self._get_session_key_part(session_key)
+                if key:
+                    return self._session_hashing_algorithm(key.encode('ascii')).hexdigest()
+            return session_key
+        return None
+
     def _get_new_session_key(self):
         "Return session key that isn't being used."
         while True:
             session_key = get_random_string(32, VALID_KEY_CHARS)
-            if not self.exists(session_key):
+            hashed_session_key = SESSION_HASHING_ALGORITHM + SESSION_KEY_DELIMITER + session_key
+            if not self.exists(session_key) and not self.exists(hashed_session_key):
```

Comment:
Is `self.exists(hashed_session_key)` necessary when `SESSION_STORE_KEY_HASH = False`?

People will expect `SESSION_STORE_KEY_HASH = False` not to incur any performance overhead.